authui/src/passkey.ts

```diff
@@ -107,21 +107,91 @@ function serializeAssertionResponse(credential: PublicKeyCredential) {
   };
 }
 
+function isSafariCancel(err: unknown) {
+  return (
+    err instanceof DOMException &&
+    err.name === "NotAllowedError" &&
+    // "This request has been cancelled by the user."
+    /cancel/i.test(err.message)
+  );
+}
+
+function isSafariTimeout(err: unknown) {
+  return (
+    err instanceof DOMException &&
+    err.name === "HierarchyRequestError" &&
+    // "The operation would yield an incorrect node tree."
+    /node tree/i.test(err.message)
+  );
+}
+
+function isSafariAndroidError(err: unknown) {
+  // This happens when the user chooses to use Android phone to scan QR code.
+  // And the passkey is not found on that Android phone.
+  // If security key is used, the error message is shown in the modal dialog and the dialog will not close.
+  // Thus we cannot detect security key error.
+  return (
+    err instanceof DOMException &&
+    err.name === "NotAllowedError" &&
+    // "Operation failed."
+    /fail/i.test(err.message)
```

Comment:
If we need to use some keywords in the message to guess the scenario. I think we should avoid using some words that are too generic (e.g. fail), seems it is easier to match some unexpected cases. I think the word `cancel` or `abort` are fine since they have some user cancel meaning. If the error message is already too general, maybe we could match a longer phrase to avoid matching some unexpected cases. What do you think?

Reply:
Updated to use longer phrase for matching.